src/plugins/discover/server/locator/lib/get_search_objects.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import {
+  injectReferences,
+  parseSearchSourceJSON,
+  SerializedSearchSourceFields,
+} from '@kbn/data-plugin/common';
+import { DeepPartial } from '@kbn/utility-types';
+import { LocatorServicesDeps, SavedSearchObjectType } from '..';
+import { DiscoverAppLocatorParams } from '../../../common';
+
+// Correction for return type of non-generic parseSearchSourceJSON
+type ParsedSearchSourceJSON = SerializedSearchSourceFields & { indexRefName?: string };
+
+function isSavedSearchObject(
+  savedSearch: SavedSearchObjectType | unknown
+): savedSearch is SavedSearchObjectType {
+  return (
+    (savedSearch as DeepPartial<SavedSearchObjectType> | undefined)?.attributes
+      ?.kibanaSavedObjectMeta?.searchSourceJSON != null
+  );
+}
+
+/**
+ * Retrieves saved search and searchSource object from discover locator params
+ * NOTE: This only supports discover content with SavedSearchObjectType. DataView-only is not supported.
```

Comment:
Question: should we support the conversions for `DiscoverAppLocatorParams` that do **not** have a `searchSourceId`? Given how this PR will be used, it is OK to wait until a future PR to support that.